### PR TITLE
Refactor to task and bake; add support for multi-platform images

### LINKED ISSUE #104: Build select images using arm64 archicecture
Currently our images are all built for amd64 which works fine for production. But for local development, folks on Apple Silicon may want to build previews from our branches on the https://github.com/TugboatQA/tugboat-test/ repository. We should build arm64 for whatever images those branches use. I believe we are already using buildx, so shouldn't be too tricky there, but we will potentially need some not insignificant refactoring to our makefile / scripts.

### PR DESCRIPTION
**Pull request by q0rban**
To test:

1. Create a .env.local
2. Add the AWS environment variables (see **Tugboat AWS S3 Access Key** in 1Password) in .env.local (see .env)
3. Set the`NAMESPACE` to your Docker hub username in .env.local
4. Run `docker login` to the account above
5. Make sure you have [`task`](https://taskfile.dev) v3.28.0 or newer installed.
6. Make sure you have GNU Parallel installed. (`brew install parallel` on macOS, `apt-get install parallel` on Debian / Ubuntu)
7. Run `PUSH=1 time task`
8. Make note of any errors you may encounter.
9. Make note of the time the command took to run.
10. Try running the command again if you encounter errors and see if it works from there.
11. Make note of the time the command took to run.
12. Run `task clean-all` to clean everything up afterward. Ensure all generated files are cleaned out, images have been removed, and `docker buildx ls` doesn't show the `tugboatqa-bakery` builder.

Resolves #104

### CODE REVIEW

**Comment by q0rban** [comment]
@serverbots test this please

**Comment by q0rban** [comment]
@serverbots test this please

**Comment by q0rban** [comment]
@serverbots test this please

**Comment by q0rban** [comment]
@serverbots test this please

**Comment by rabbitlair** [comment]
I get the following error when running step 6. The error references a different image each time. Also, when running the command, the password to decrypt the Docker authentication is asked several times.

```
------
 > [varnish-stable internal] load metadata for docker.io/library/varnish:stable:
------
------
 > [percona-8-0-33-25-centos internal] load metadata for docker.io/library/percona:8.0.33-25-centos:
------
Dockerfile:1
--------------------
   1 | >>> FROM varnish:stable
   2 |     
   3 |     USER root
--------------------
ERROR: failed to solve: varnish:stable: error getting credentials - err: exit status 1, out: `error getting credentials - err: exit status 1, out: `exit status 2: gpg: descifrado fallido: No tenemos la clave secreta``
7 v0.12.1 buildkitd
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryClient.func1.1.1
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:72
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryClientInterceptor.func1
	/src/vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:99
github.com/moby/buildkit/session.filterClient.func1
	/src/session/session.go:186
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryClient.func1.1.1
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:72
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryClient.func1
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:81
google.golang.org/grpc.(*ClientConn).Invoke
	/src/vendor/google.golang.org/grpc/call.go:35
github.com/moby/buildkit/session/auth.(*authClient).GetTokenAuthority
	/src/session/auth/auth.pb.go:878
github.com/moby/buildkit/session/auth.GetTokenAuthority.func1
	/src/session/auth/auth.go:107
github.com/moby/buildkit/session.(*Manager).Any
	/src/session/group.go:82
github.com/moby/buildkit/session/auth.GetTokenAuthority
	/src/session/auth/auth.go:104
github.com/moby/buildkit/util/resolver.(*dockerAuthorizer).AddResponses
	/src/util/resolver/authorizer.go:184
github.com/containerd/containerd/remotes/docker.(*request).retryRequest
	/src/vendor/github.com/containerd/containerd/remotes/docker/resolver.go:627
github.com/containerd/containerd/remotes/docker.(*request).doWithRetries
	/src/vendor/github.com/containerd/containerd/remotes/docker/resolver.go:606
github.com/containerd/containerd/remotes/docker.(*dockerResolver).Resolve
	/src/vendor/github.com/containerd/containerd/remotes/docker/resolver.go:289
github.com/moby/buildkit/util/resolver.(*Resolver).Resolve
	/src/util/resolver/pool.go:204
github.com/moby/buildkit/util/imageutil.Config
	/src/util/imageutil/config.go:142
github.com/moby/buildkit/source/containerimage.(*Source).ResolveImageConfig.func1
	/src/source/containerimage/pull.go:115
github.com/moby/buildkit/util/flightcontrol.(*call[...]).run
	/src/util/flightcontrol/flightcontrol.go:121
sync.(*Once).doSlow
	/usr/local/go/src/sync/once.go:74
sync.(*Once).Do
	/usr/local/go/src/sync/once.go:65
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1598

7 v0.12.1 buildkitd
github.com/moby/buildkit/frontend/dockerfile/dockerfile2llb.toDispatchState.func3.1
	/src/frontend/dockerfile/dockerfile2llb/convert.go:413
golang.org/x/sync/errgroup.(*Group).Go.func1
	/src/vendor/golang.org/x/sync/errgroup/errgroup.go:75
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1598

323161 v0.11.2-desktop.1 /usr/lib/docker/cli-plugins/docker-buildx buildx bake --builder tugboatqa-bakery --file ./bake.json
google.golang.org/grpc.getChainUnaryInvoker.func1
	google.golang.org/grpc@v1.53.0/clientconn.go:361
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryClientInterceptor.func1
	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.40.0/interceptor.go:99
github.com/moby/buildkit/client.filterInterceptor.func1
	github.com/moby/buildkit@v0.12.0/client/client.go:384
google.golang.org/grpc.chainUnaryClientInterceptors.func1
	google.golang.org/grpc@v1.53.0/clientconn.go:349
google.golang.org/grpc.(*ClientConn).Invoke
	google.golang.org/grpc@v1.53.0/call.go:35
github.com/moby/buildkit/frontend/gateway/pb.(*lLBBridgeClient).Solve
	github.com/moby/buildkit@v0.12.0/frontend/gateway/pb/gateway.pb.go:2876
github.com/moby/buildkit/client.(*gatewayClientForBuild).Solve
	github.com/moby/buildkit@v0.12.0/client/build.go:89
github.com/moby/buildkit/frontend/gateway/grpcclient.(*grpcClient).Solve
	github.com/moby/buildkit@v0.12.0/frontend/gateway/grpcclient/client.go:411
github.com/docker/buildx/build.BuildWithResultHandler.func3.1.2
	github.com/docker/buildx/build/build.go:935
github.com/moby/buildkit/frontend/gateway/grpcclient.(*grpcClient).Run
	github.com/moby/buildkit@v0.12.0/frontend/gateway/grpcclient/client.go:214
github.com/moby/buildkit/client.(*Client).Build.func2
	github.com/moby/buildkit@v0.12.0/client/build.go:59
github.com/moby/buildkit/client.(*Client).solve.func3
	github.com/moby/buildkit@v0.12.0/client/solve.go:283
golang.org/x/sync/errgroup.(*Group).Go.func1
	golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75
runtime.goexit
	runtime/asm_amd64.s:1598

7 v0.12.1 buildkitd
github.com/moby/buildkit/frontend/gateway.(*llbBridgeForwarder).Return
	/src/frontend/gateway/gateway.go:917
github.com/moby/buildkit/control/gateway.(*GatewayForwarder).Return
	/src/control/gateway/gateway.go:135
github.com/moby/buildkit/frontend/gateway/pb._LLBBridge_Return_Handler.func1
	/src/frontend/gateway/pb/gateway.pb.go:3216
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryServerInterceptor.func1
	/src/vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/interceptor.go:342
main.unaryInterceptor.func1
	/src/cmd/buildkitd/main.go:603
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1.1
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:25
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1
	/src/vendor/github.com/grpc-ecosystem/go-grpc-middleware/chain.go:34
github.com/moby/buildkit/frontend/gateway/pb._LLBBridge_Return_Handler
	/src/frontend/gateway/pb/gateway.pb.go:3218
google.golang.org/grpc.(*Server).processUnaryRPC
	/src/vendor/google.golang.org/grpc/server.go:1336
google.golang.org/grpc.(*Server).handleStream
	/src/vendor/google.golang.org/grpc/server.go:1704
google.golang.org/grpc.(*Server).serveStreams.func1.2
	/src/vendor/google.golang.org/grpc/server.go:965
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1598

323161 v0.11.2-desktop.1 /usr/lib/docker/cli-plugins/docker-buildx buildx bake --builder tugboatqa-bakery --file ./bake.json
google.golang.org/grpc.getChainUnaryInvoker.func1
	google.golang.org/grpc@v1.53.0/clientconn.go:361
go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.UnaryClientInterceptor.func1
	go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc@v0.40.0/interceptor.go:99
github.com/moby/buildkit/client.filterInterceptor.func1
	github.com/moby/buildkit@v0.12.0/client/client.go:384
google.golang.org/grpc.chainUnaryClientInterceptors.func1
	google.golang.org/grpc@v1.53.0/clientconn.go:349
google.golang.org/grpc.(*ClientConn).Invoke
	google.golang.org/grpc@v1.53.0/call.go:35
github.com/moby/buildkit/api/services/control.(*controlClient).Solve
	github.com/moby/buildkit@v0.12.0/api/services/control/control.pb.go:2208
github.com/moby/buildkit/client.(*Client).solve.func2
	github.com/moby/buildkit@v0.12.0/client/solve.go:258
golang.org/x/sync/errgroup.(*Group).Go.func1
	golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75
runtime.goexit
	runtime/asm_amd64.s:1598

323161 v0.11.2-desktop.1 /usr/lib/docker/cli-plugins/docker-buildx buildx bake --builder tugboatqa-bakery --file ./bake.json
github.com/moby/buildkit/client.(*Client).solve.func2
	github.com/moby/buildkit@v0.12.0/client/solve.go:273
golang.org/x/sync/errgroup.(*Group).Go.func1
	golang.org/x/sync@v0.2.0/errgroup/errgroup.go:75
runtime.goexit
	runtime/asm_amd64.s:1598

task: Failed to run task "default": exit status 1
Command exited with non-zero status 201
24.95user 13.49system 0:28.94elapsed 132%CPU (0avgtext+0avgdata 109532maxresident)k
0inputs+51560outputs (264major+2212946minor)pagefaults 0swaps
```

**Comment by q0rban** [comment]
@rabbitlair I think maybe I left out a step to run `docker login` prior to running?

**Comment by rabbitlair** [comment]
@q0rban, even when logging in to Docker, the issue is still happening to me :(

**Comment by q0rban** [comment]
Oh, perhaps docker is using gpg to decrypt the docker credentials, and so maybe you need to get that working again?

**Comment by q0rban** [comment]
@rabbitlair can you investigate to see if there's an issue with docker login and gpg on your machine?

**Comment by rabbitlair** [comment]
I have been researching, and, indeed, it seems like there is some kind of conflict with my Docker authentication. Docker Desktop is able to authenticate properly, same as Docker CLI, but the issue is still happening to me. I will try to test from a different workstation

**Comment by q0rban** [comment]
@rabbitlair I'd like to bump the priority on this as this PR fixes some issues that is causing the current images build to fail on Jenkins.

**Comment by rabbitlair** [comment]
Tested and verified, it works well :100: 

The job was interrupted like 3 or 4 times, but once it was run again, it continued from the spot it was before stopping. The errors I have seen are the following:

Connectivity
```
ERROR: failed to receive status: rpc error: code = Unavailable desc = error reading from server: EOF
363 v0.11.2 /usr/libexec/docker/cli-plugins/docker-buildx buildx bake --builder tugboatqa-bakery --file ./bake.json
```

Docker error
```
ERROR: failed to solve: failed commit on ref "layer-sha256:0b5ede2a1ee857a549f04a8ac45158b20e57f5e5d9634c67e0c6a2e1e87817c9": rename /var/lib/buildkit/runc-overlayfs/content/ingest/73cb3ba62418b95ba958f5fea9da8aee7343987ccfbe9a9459c83fc0af11cd90/data /var/lib/buildkit/runc-overlayfs/content/blobs/sha256/0b5ede2a1ee857a549f04a8ac45158b20e57f5e5d9634c67e0c6a2e1e87817c9: no such file or directory
```

Error when pushing images
```
images/mongo/7.0.0-jammy/image.tar  devmapper: Thin Pool has 2770 free data blocks which is less than minimum required 3890 free data blocks. Create more free space in thin pool or use dm.min_free_space option to change behavior
```

**Review by rabbitlair — APPROVED** [review]

**Comment by q0rban** [comment]
Awesome, thank you @rabbitlair ! Let's merge this in and work on improvements in follow-ups.

File: Taskfile.yml [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+# The following are named snippets to avoid duplication.
+require_jq: &require_jq
+  sh: 'command -v jq'
+  msg: 'You must install jq.'
+require_parallel: &require_parallel
+  sh: 'command -v parallel'
+  msg: 'You must install GNU Parallel.'
+require_v28: &require_v28
+  sh: 'test {{.TASK_MINOR_VERSION}} -ge 28'
+  msg: 'Please upgrade Task to v3.28.0 or greater.'
```
Comment (deviantintegral) [inline Taskfile.yml:13]: @justafish this pattern may be useful in drainpipe.

File: Taskfile.yml [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      - grep -q "max-parallelism = ${BUILDX_MAX_PARALLELISM}\$" buildkitd.toml
+    sources:
+      - templates/buildkitd.template.toml
+    generates:
+      - ./buildkitd.toml
+    cmds:
+      - envsubst < templates/buildkitd.template.toml > buildkitd.toml
+
+  bake:
+    desc: '[subtask] Runs docker buildx bake using the bake.json generated by bakefile.'
```
Comment (deviantintegral) [inline Taskfile.yml:114]: @q0rban @rabbitlair bake seems like good show and tell material!

**Comment by deviantintegral** [comment]
I just wanted to say how excited I am to have multi-platform images 👏 👏 👏 .

**Pull request merged**